presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerClusterStatusSender.java

```diff
@@ -0,0 +1,142 @@
+        this.statusSupplier = statusResource::getStatus;
+        this.executor = requireNonNull(executor, "executor is null");
+        this.resourceManagerConfig = requireNonNull(resourceManagerConfig, "resourceManagerConfig is null");
+    }
+
+    public ResourceManagerClusterStatusSender(
+            DriftClient<ResourceManagerClient> resourceManagerClient,
+            InternalNodeManager internalNodeManager,
+            Supplier<NodeStatus> statusResource,
+            ScheduledExecutorService executor,
+            ResourceManagerConfig resourceManagerConfig)
+    {
+        this.resourceManagerClient = requireNonNull(resourceManagerClient, "resourceManagerService is null");
+        this.internalNodeManager = requireNonNull(internalNodeManager, "internalNodeManager is null");
+        this.statusSupplier = requireNonNull(statusResource, "statusResource is null");
+        this.executor = requireNonNull(executor, "executor is null");
+        this.resourceManagerConfig = requireNonNull(resourceManagerConfig, "resourceManagerConfig is null");
+    }
+
+    @PostConstruct
+    public void init()
+    {
+        nodeHeartbeatSender = new PeriodicTaskExecutor(resourceManagerConfig.getNodeHeartbeatInterval().toMillis(), executor, this::sendNodeHeartbeat);
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        queries.values().forEach(PeriodicTaskExecutor::stop);
+        if (nodeHeartbeatSender != null) {
+            nodeHeartbeatSender.stop();
+        }
+    }
+
+    @Override
+    public void registerQuery(ManagedQueryExecution queryExecution)
+    {
+        QueryId queryId = queryExecution.getBasicQueryInfo().getQueryId();
+        queries.computeIfAbsent(queryId, unused -> new PeriodicTaskExecutor(
+                resourceManagerConfig.getQueryHeartbeatInterval().toMillis(),
+                executor,
+                () -> sendQueryHeartbeat(queryExecution)));
+        queryExecution.addStateChangeListener(newState -> {
+            if (newState.isDone()) {
+                queries.computeIfPresent(queryId, (unused, queryHeartbeatSender) -> {
+                    queryHeartbeatSender.forceRun();
+                    queryHeartbeatSender.stop();
+                    return null;
+                });
+            }
+        });
+    }
+
+    private void sendQueryHeartbeat(ManagedQueryExecution queryExecution)
+    {
+        BasicQueryInfo basicQueryInfo = queryExecution.getBasicQueryInfo();
+        String nodeIdentifier = internalNodeManager.getCurrentNode().getNodeIdentifier();
+        getResourceManagers().forEach(hostAndPort ->
+                resourceManagerClient.get(Optional.of(hostAndPort.toString())).queryHeartbeat(nodeIdentifier, basicQueryInfo));
+    }
+
+    private void sendNodeHeartbeat()
+    {
+        getResourceManagers().forEach(hostAndPort ->
+                resourceManagerClient.get(Optional.of(hostAndPort.toString())).nodeHeartbeat(statusSupplier.get()));
```

Comment:
But in that case why do we still want to pass the `addressSelectionContext`  ? (since `RandomResourceManagerAddressSelector` anyway ignores `addressSelectionContext` ?) 😄

Reply:
This is fixed now, the address selector should short circuit and just use the supplied context whenever it's present.